### ISSUE TITLE
prune flag --keep-after for keeping snapshots relative to now

### ISSUE THREAD

**Issue by phillipp** [restic/restic#3392]
We have a backup system implemented with restic where we need to prune snapshots based on a date relative to the current time. We keep snapshots from the same data in multiple repositories due to the repos getting too large (tens of thousands of snapshots per repo).

We tried pruning these by parsing the snapshot list and pruning with a list of snapshot IDs, but prune lists the snapshots from source for every ID given because of how the snapshot list building works (giving the prune command 1000 snapshot IDs pulls the snapshot list from the REST backend 1000 times).

We then added a `--keep-after` flag that works like `--keep-within`, but with the current time rather than the time of the last snapshot.

Would you be interested in a pull request for that?

**Comment by aawsome** [comment]
> prune lists the snapshots from source for every ID given because of how the snapshot list building works (giving the prune command 1000 snapshot IDs pulls the snapshot list from the REST backend 1000 times).

I also noticed this and would say we could claim this as performace bug.
For a list of (short) snapshot IDs it must be sufficient to list the snapshot files only once from the repository.
In fact, I already started reworking and unifying the snapshot processing (also implementing ideas like 'latest-1' as in #1965) but did run a bit out of time to finish it.

However, in any case we can simply remove the backend listing when a long ID is given. If you call restic from a script it should be no problem to use long snapshot IDs.
I started #3395, feel free to try it out.

**Comment by MichaelEischer** [comment]
> In fact, I already started reworking and unifying the snapshot processing (also implementing ideas like 'latest-1' as in #1965) but did run a bit out of time to finish it.

I agree list should only be called once at the backend, but please fix the performance issue separately from adding new features.

**Comment by phillipp** [comment]
@MichaelEischer @aawsome 
I agree, a fix for the performance problem would be good - but the question remains if I should open a pull request for `--keep-after` or not? ;-)

**Comment by MichaelEischer** [comment]
@phillipp What exactly prevents you from using `--keep-within`? One important feature of the `--keep-*` flags is that these prevent accidental deletions of all snapshots within a snapshot group.

**Comment by phillipp** [comment]
@MichaelEischer It won't delete snapshots that have no recent backups. In our case, we switch repositories for a data directory when the snapshot count in a repo reaches a certain limit (like a few ten thousands) because of performance problems. We have around 18 million snapshots right now.

**Comment by MichaelEischer** [comment]
Not deleting snapshots without recent backups is in fact intentional as this makes it harder to accidentally remove your backups.

See https://restic.readthedocs.io/en/latest/060_forget.html#removing-snapshots-according-to-a-policy :
> For safety reasons, restic refuses to act on an “empty” policy. For example, if one were to specify --keep-last 0 to forget all snapshots in the repository, restic will respond that no snapshots will be removed. To delete all snapshots, use --keep-last 1 and then finally remove the last snapshot ID manually (by passing the ID to forget).

**Comment by phillipp** [comment]
@MichaelEischer That is understood. However, *we* **don't** want that safety.
I just wanted to know if we should open a pull request for the (non-safety) feature/flag we added. I guess it just does not fit the philosophy of restic, so the answer is "no". Feel free to simply close the issue if you agree.

**Comment by MichaelEischer** [comment]
Indeed, the answer is no. I'll create a separate issue for the performance issue.
